### DISCUSSION TITLE
Configure and Initialize Python.NET in a C# Console App on Linux containerized with Docker

### DISCUSSION THREAD

**Discussion by jimnoble** [pythonnet/pythonnet#2296] [Q&A]
### Environment

-   Pythonnet version: 3.0.3
-   Python version: 3.12
-   Operating System: Linux
-   .NET Runtime: .NET 7.0

### Details

I am trying to run a C#/.NET console application that uses the Python.NET library to interpret Python scripts at runtime. I want this app to run on Linux with a Docker container.

The problem I am having is that Python.NET requires you to initialize the `Runtime.PythonDLL` property with the path to the Python library binary.

The answer on [this post](https://stackoverflow.com/questions/69670715/how-to-set-runtime-pythondll-for-pythonnet-for-console-wpf-net-application-p) makes me suspect I am looking for a file named e.g. libpython3.12.so.

I have created a dockerfile (see below) that installs the `python:3` image to a virtual environment and copies the result into my final environment. But I do not see any file named like `libpython3.12.so`. If I try to provide the `/app/python/bin/python` executable (which _does_ exist), Python.NET throws an exception (see below) that the file does not exist.

I can't find an explicit guide to how this should be done, does one exist?

How should I configure my dockerfile, and what is the path to the `libpython*.so` file that I should assign to `Runtime.PythonDLL`?

the `.dockerfile`:
```
FROM mcr.microsoft.com/dotnet/aspnet:7.0 AS base
WORKDIR /app
EXPOSE 80
EXPOSE 443

FROM mcr.microsoft.com/dotnet/sdk:7.0 AS build
WORKDIR /src
COPY ["Farpoint.Service/Farpoint.Service.csproj", "Farpoint.Service/"]
RUN dotnet restore "Farpoint.Service/Farpoint.Service.csproj"
COPY . .
WORKDIR "/src/Farpoint.Service"
RUN dotnet build "Farpoint.Service.csproj" -c Release -o /app/build

# BEGIN Prepare python env
FROM python:3 as python
RUN python -m venv /venv
COPY --from=build /app/build/python/requirements.txt .
RUN /venv/bin/python -m pip install -r requirements.txt
# END Prepare python env

FROM build AS publish
RUN dotnet publish "Farpoint.Service.csproj" -c Release -o /app/publish /p:UseAppHost=false

FROM base AS final
WORKDIR /app
COPY --from=publish /app/publish .

# BEGIN Copy python env
COPY --from=python /venv /app/python
# END Copy python env

ENTRYPOINT ["dotnet", "Farpoint.Service.dll"]
```

The exception:
> DllNotFoundException: Could not load /app/python/bin/python with flags RTLD_NOW | RTLD_GLOBAL: /app/python/bin/python: cannot open shared object file: No such file or directory

**Comment by gerneio** [comment]
@jimnoble Any luck?

In my case I'm running under a windows container `mcr.microsoft.com/dotnet/runtime:8.0.1-windowsservercore-ltsc2019` and using an embedded python runtime that I include in my projects output folder ([see](https://github.com/pythonnet/pythonnet/issues/467#issuecomment-1912271678)).

The exception:

```
Unhandled exception. System.TypeInitializationException: The type initializer for 'Delegates' threw an exception.
 ---> System.DllNotFoundException: Could not load C:\app\python\python312.dll.
 ---> System.ComponentModel.Win32Exception (126): The specified module could not be found.
   --- End of inner exception stack trace ---
   at Python.Runtime.Platform.WindowsLoader.Load(String dllToLoad)
   at Python.Runtime.Runtime.Delegates.GetUnmanagedDll(String libraryName)
   at Python.Runtime.Runtime.Delegates..cctor()
   --- End of inner exception stack trace ---
   at Python.Runtime.Runtime.Delegates.get_Py_SetPythonHome()
   at Python.Runtime.Runtime.Py_SetPythonHome(IntPtr home)
   at Python.Runtime.PythonEngine.<>c.<set_PythonHome>b__27_0()
   at Python.Runtime.Runtime.<>c__DisplayClass106_0.<TryUsingDll>b__0()
   at Python.Runtime.Runtime.TryUsingDll(Action op)
   at Python.Runtime.PythonEngine.set_PythonHome(String value)
   at TestPythonNet.Python.PythonEmailProxy.PythonStatup(ILogger logger)
```

Can confirm that that the DLL does exists at `C:\app\python\python312.dll`. Same program works fine on local windows dev machine (i.e. outside container).

FYI, this is the code I'm using to set `PythonDLL` and `PythonHome` (which could vary per platform I suppose):

```csharp
var pythonDir = Path.Combine(Environment.CurrentDirectory, @"python");

Runtime.PythonDLL = Path.Combine(pythonDir, "python312.dll");
PythonEngine.PythonHome = pythonDir;

PythonEngine.Initialize();
```

**Reply by gerneio** [reply]
Was able to get mine working, albeit by bloating my container size probably. For my situation, ultimately I found that when `PythonNet` tries to load the DLL set by by `Runtime.PythonDLL`, it calls into a win32 interop lib (kernel32) to call a `LoadLibrary` function, which returns an empty pointer. The `Win32Exception` I shared in my last post (error code = 126) stated that "[The specified module could not be found](https://learn.microsoft.com/en-us/windows/win32/debug/system-error-codes--0-499-#:~:text=no%20volume%20label.-,ERROR_MOD_NOT_FOUND,-126%20(0x7E))" which can indicate a [number of things](https://stackoverflow.com/a/26231466/10388359):

> DLL is not there on specified path or
> DLLs is not supported with platform or
> dependent dlls are not loaded before you build your native DLL

I created a small powershell script to quickly test this precise issue when loading the `python312.dll`, so for the most part excluding `PythonNet` out of the reproducible steps.

`.\TestLoadingDLL.ps1 "C:\python\python312.dll"`
```ps
$dllToLoad = $args[0]

if ($dllToLoad -eq "" -or !(Test-Path $dllToLoad)) {
    throw "File doesn't exist: $dllToLoad"
}

$MethodDefinition = @"
private const string NativeDll = "kernel32.dll";

[DllImport(NativeDll, SetLastError = true)]
public static extern IntPtr LoadLibrary(string dllToLoad);
"@

$Kernel32 = Add-Type -MemberDefinition $MethodDefinition -Name 'Kernel32' -Namespace 'Win32' -PassThru

$ptr = $Kernel32::LoadLibrary($dllToLoad);
$lastError = [System.Runtime.InteropServices.Marshal]::GetLastWin32Error()

Write-Host $ptr

if ($ptr -eq 0)
{
    Write-Error $lastError

    throw new-object System.ComponentModel.Win32Exception -ArgumentList $lastError
}
```

Out of curiosity, I loaded up a windows container that already had python installed ([python:3.12.1-windowsservercore-1809](https://github.com/docker-library/python/blob/master/3.12/windows/windowsservercore-1809/Dockerfile)), and ran the above powershell script against the python dll in it's default install directory, and it worked no issue. So probably what is happening here is that the python installer that runs during the building of that python docker image ([shown here](https://github.com/docker-library/python/blob/f064da21d58370cebb45e89d17f79a290f1d9086/3.12/windows/windowsservercore-1809/Dockerfile#L16-L34)), installs some system level dependencies that are required when loading that python DLL, which are simply just missing when I'm using the embedded python runtime. Looking through the python docs under this [embedded section](https://docs.python.org/3/using/windows.html#the-embeddable-package), it references that you must make sure to have the `Microsoft C Runtime` installed. It also states that if the `ucrtbase.dll` exists in the system directory, then it is already installed, which it was for the initial servercore image I tried, so unsure if it was related. Running `python.exe` under either image worked w/o issue, so this was a specific issue to loading that DLL via that `Kernel32::LoadLibrary()` function.

Ultimately my problems stemmed from utilizing the embed python runtime as I mentioned in my last post, so what I ended up doing was using the python docker image as my base and installing the dotnet runtime via this [dotnet-install script](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-install-script). This allowed me to move forward with actually getting `PythonNet` to start. There were several other headaches I had to deal with such as modifying the `PATH` environment var for `dotnet`, as well as figuring out how and what to set `PythonEngine.PythonPath` (it didn't behave the same as my local dev machine). But after several hours of tweaking I did finally get it all working (basically have the python runtime installed twice though).

`Dockerfile`
```dockerfile
FROM python:3.12.1-windowsservercore-1809

SHELL ["powershell", "-Command", "$ErrorActionPreference = 'Stop'; $ProgressPreference = 'SilentlyContinue';"]

RUN &powershell -NoProfile -ExecutionPolicy unrestricted -Command "[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12; &([scriptblock]::Create((Invoke-WebRequest -UseBasicParsing 'https://dot.net/v1/dotnet-install.ps1'))) -Runtime dotnet -Version 8.0.0"

RUN [ "powershell", "-command", "[Environment]::SetEnvironmentVariable(\"Path\", $env:Path + \";C:\\Users\\ContainerAdministrator\\AppData\\Local\\Microsoft\\dotnet\", \"Machine\")" ]

# Set the working directory
WORKDIR /app

ADD MyProject /app/

ENTRYPOINT ["dotnet", "MyProject.dll"]
```

`C# PythonNet Startup`
```cs
var pythonDir = Path.Combine(Environment.CurrentDirectory, @"python");

// hardcoded for now, but can probably be made to a bit more dynamic if we search through the `python` output directory
var pyVer = "312";

Runtime.PythonDLL = Path.Combine(pythonDir, $"python{pyVer}.dll");
PythonEngine.PythonHome = pythonDir;

// Mimicking what was in `sys.path` when running on local dev machine (different when running under docker container for some reason)
string[] dirs = [
    $"python\\python{pyVer}.zip",
    "python\\DLLs",
    "python\\Lib",
    "",
    "python",
    "python\\Lib\\site-packages"
];

var pythonPaths = string.Join(';', dirs.Select(e => Path.Combine(Environment.CurrentDirectory, e)));

PythonEngine.PythonPath = pythonPaths;

PythonEngine.Initialize();
```

This is far from the most optimal way to resolve this, but it gets the job done for now.

Sorry @jimnoble, I don't think any of this applies to your situation since my issue seemed like a pretty windows specific issue. I initially thought our two issues were identical due to the `DllNotFoundException`, however after digging into it, it was really the `Win32Exception` that was the main indicator on my issue. For your issue, if you're able to paste the full stack trace, then perhaps we can reason through to a solution.

**Reply by RekingLessen** [reply]
I have the same issue with the PYTHONNET_PYDLL path setup. I've installed the python3.11. The pythonnet works well in windows after I specified the dll path to C:\Users\103296\AppData\Local\Programs\Python\Python311\Python311.dll in the system environment variables.

I tried to find the equivalent library file in docker environment. As the document mentioned,  I found the libpython3.11.so in the folder begin with "config", I've specified the PYTHONNET_PYDLL environment variable in the docker file.
ENV PYTHONNET_PYDLL=/usr/lib/python3.11/config-3.11-x86_64-linux-gnu/libpython3.11.so.

Unfortunately, the same error still occurred.

Unhandled exception. Python.Runtime.BadPythonDllException: Runtime.PythonDLL was not set or does not point to a supported Python runtime DLL. See https://github.com/pythonnet/pythonnet#embedding-python-in-net
2024-02-02T21:02:36.526974682Z  ---> System.MissingMethodException: Failed to load symbol Py_IncRef: dotnet: undefined symbol: Py_IncRef
2024-02-02T21:02:36.526980888Z    at Python.Runtime.Platform.PosixLoader.GetFunction(IntPtr dllHandle, String name)
2024-02-02T21:02:36.526984945Z    at Python.Runtime.Runtime.Delegates.GetFunctionByName(String functionName, IntPtr libraryHandle)
2024-02-02T21:02:36.526988900Z    --- End of inner exception stack trace ---
2024-02-02T21:02:36.526992497Z    at Python.Runtime.Runtime.TryUsingDll[T](Func`1 op)

I really appreciate if someone can provide any help.

**Comment by giodynet** [comment]
** With miniconda 3 and python 3.12. Dockerfile:

FROM mcr.microsoft.com/dotnet/runtime:6.0 AS base
ENV PATH="/root/miniconda3/bin:$PATH"
ENV PATH="/app:$PATH"
ENV PATH="/root/miniconda3/lib/:$PATH"
ARG PATH="/root/miniconda3/bin:$PATH"
RUN apt-get update && apt upgrade -y
RUN apt-get install software-properties-common -y
RUN apt-get install -y wget && rm -rf /var/lib/apt/lists/*
RUN wget \
    https://repo.anaconda.com/miniconda/Miniconda3-py312_24.3.0-0-Linux-x86_64.sh \
    && mkdir /root/.conda \
    && bash Miniconda3-py312_24.3.0-0-Linux-x86_64.sh  -b \
    && rm -f Miniconda3-py312_24.3.0-0-Linux-x86_64.sh 
RUN pip install -Iv --no-input pygrib==2.1.5

FROM mcr.microsoft.com/dotnet/sdk:6.0 AS build
ARG BUILD_CONFIGURATION=Release
...
..
etc etc

** in the program.cs

 Environment.SetEnvironmentVariable("PYTHONHOME", "/root/miniconda3/lib");
 Environment.SetEnvironmentVariable("PYTHONNET_PYDLL", "/root/miniconda3/lib/libpython3.12.so");
 if (!PythonEngine.IsInitialized)
 { 
     PythonEngine.Initialize();
 }

it takes me 2 days but it's working..
Cheers
Giody
